### PR TITLE
ROC-3736 Pay hub with CCD reference

### PR DESCRIPTION
**Pull request by timja-kainos**

### CODE REVIEW

File: src/test/http-mocks/claim-store.ts [hunk 1]
```diff
@@ -94,7 +94,7 @@ export function resolveRetrieveClaimByClaimNumber (claimOverride?: object): mock
 
 export function rejectRetrieveClaimByExternalIdWithNotFound (reason: string) {
   mock(`${serviceBaseURL}/claims`)
-    .get(new RegExp('/[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}'))
+    .get(new RegExp('/[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'))
```
Comment (yarpo) [inline src/test/http-mocks/claim-store.ts]: why not `^` at the beginning of the regEx?

Comment (timja-kainos) [reply src/test/http-mocks/claim-store.ts]: Only worried about terminating it, it was matching the wrong one in route tests

File: src/test/features/claim/routes/pay-by-account.ts [hunk 2]
```diff
@@ -75,7 +75,11 @@ describe('Claim : Pay by Fee Account page', () => {
 
     it('should return 500 and render error page when form is valid and cannot save draft', async () => {
```
Comment (geek-soft) [inline src/test/features/claim/routes/pay-by-account.ts]: @timja-kainos please add route test for failure in retrieving payment reference.

File: src/test/features/claim/routes/pay-by-account.ts [hunk 3]
```diff
@@ -75,7 +75,11 @@ describe('Claim : Pay by Fee Account page', () => {
...
       idamServiceMock.resolveRetrieveUserFor('1', ...roles)
+      idamServiceMock.resolveRetrieveServiceToken()
       draftStoreServiceMock.rejectSave(100, 'HTTP error')
+      feesServiceMock.resolveCalculateIssueFee()
```
Comment (geek-soft) [inline src/test/features/claim/routes/pay-by-account.ts]: I guess case was failing with 500 before which was caused by missing mocks.

Comment (timja-kainos) [reply src/test/features/claim/routes/pay-by-account.ts]: this file had soo many connection refused happening and not doing what the tests description was.
I fixed that

File: src/test/features/claim/routes/pay-by-account.ts (hunk 2 above)
Comment (timja-kainos) [reply src/test/features/claim/routes/pay-by-account.ts]: done

File: src/test/features/claim/routes/pay-by-account.ts [hunk 4]
```diff
@@ -103,11 +108,29 @@ describe('Claim : Pay by Fee Account page', () => {
           .toLocation(ClaimPaths.claimSubmittedPage.uri.replace(':externalId', claimStoreServiceMock.sampleClaimObj.externalId)))
     })
 
+    it('should return 500 and render error page when form is valid and cannot retrieve case reference', async () => {
+      idamServiceMock.resolveRetrieveUserFor('1', ...roles)
+      draftStoreServiceMock.resolveUpdate()
+      feesServiceMock.resolveCalculateIssueFee()
+      claimStoreServiceMock.resolveRetrieveClaimByExternalId()
+      claimStoreServiceMock.rejectRetrievePaymentReference()
+      feesServiceMock.resolveCalculateIssueFee()
```
Comment (geek-soft) [inline src/test/features/claim/routes/pay-by-account.ts]: Do we have two calls to fees registry to calculate fee?

File: src/test/features/claim/routes/pay-by-account.ts [hunk 5]
```diff
@@ -103,14 +107,30 @@ describe('Claim : Pay by Fee Account page', () => {
           .toLocation(ClaimPaths.claimSubmittedPage.uri.replace(':externalId', claimStoreServiceMock.sampleClaimObj.externalId)))
     })
 
+    it('should return 500 and render error page when form is valid and cannot retrieve case reference', async () => {
+      idamServiceMock.resolveRetrieveUserFor('1', ...roles)
+      draftStoreServiceMock.resolveUpdate()
+      feesServiceMock.resolveCalculateIssueFee()
+      claimStoreServiceMock.resolveRetrieveClaimByExternalId()
```
Comment (geek-soft) [inline src/test/features/claim/routes/pay-by-account.ts]: That seems to be unneeded as retrieval should happen only when payment was successfully initiated. The same with `draftStoreServiceMock.resolveUpdate()` that should happen only when payment was successfully created.

File: src/main/features/claim/routes/pay-by-account.ts [hunk 6]
```diff
@@ -122,9 +124,9 @@ export default express.Router()
         } else {
           const payClient: PayClient = await getPayClient()
           const paymentResponse: PaymentResponse = await payClient.create(
-            res.locals.user,
+            user,
             draft.document.feeAccount.reference,
-            draft.document.externalId,
+            await ClaimStoreClient.retrievePaymentReference(draft.document.externalId, user),
```
Comment (geek-soft) [inline src/main/features/claim/routes/pay-by-account.ts]: That seems alright under condition that `ClaimStoreClient.retrievePaymentReference` is idempotent and can be called many times. If such condition is violated then users would not be able to recover when draft updated failed.

Comment (timja-kainos) [reply src/main/features/claim/routes/pay-by-account.ts]: it is idempotent given the first call finished, I wouldn't recommend calling it 3 times at the same time though

**Pull request merged**